sr_speech_control/src/sr_speech_control/speech_control.py

```diff
@@ -15,66 +15,46 @@
 # with this program. If not, see <http://www.gnu.org/licenses/>.
 
 import rospy
-import time
 import speech_recognition as sr
-from difflib import get_close_matches
 from std_msgs.msg import String
 
 
 class SpeechControl(object):
-    def __init__(self, trigger_word, command_words, command_topic='speech_control',
-                 non_speaking_duration=0.1, pause_threshold=0.2):
+    def __init__(self, trigger_word, command_topic='sr_speech_control'):
+        self.microphone = sr.Microphone()
+        for i, microphone_name in enumerate(sr.Microphone.list_microphone_names()):
+            if "pulse" in microphone_name:
+                self.microphone = sr.Microphone(device_index=i)
+                rospy.loginfo("Using microphone: {}".format(microphone_name))
         self.trigger_word = trigger_word
-        self.command_words = command_words
         self.recognizer = sr.Recognizer()
-        self.microphone = sr.Microphone()
         self.command_publisher = rospy.Publisher(command_topic, String, queue_size=1)
-        self.command_to_be_executed = None
-
-        self._init_recognizer(non_speaking_duration, pause_threshold)
         self._stop_listening = self.recognizer.listen_in_background(self.microphone, self._recognizer_callback)
 
-    def _init_recognizer(self, non_speaking_duration, pause_threshold):
-        with self.microphone as source:
-            self.recognizer.adjust_for_ambient_noise(source)
-        self.recognizer.non_speaking_duration = non_speaking_duration
-        self.recognizer.pause_threshold = pause_threshold
-
     def _recognizer_callback(self, recognizer, audio):
         try:
             result = recognizer.recognize_google(audio)
+            rospy.loginfo("Recognition result: {}".format(result))
         except sr.UnknownValueError:
             return
         except sr.RequestError as e:
             rospy.logwarn("Could not request results from Google Speech Recognition service: {}".format(e))
             return
 
-        result = [str(x).lower() for x in result.split(' ')]
-        if self._filter_word(result[0], self.trigger_word) == self.trigger_word:
-            command = self._filter_word(result[1], self.command_words)
-            if command in self.command_words:
-                self.command_to_be_executed = command
-
-    def _filter_word(self, word, dictionary, offset=0.5):
-        result = get_close_matches(word, dictionary, 1, offset)
-        if not result:
-            return word
-        return result[0]
+        result = result.lower()
+        if result.startswith(self.trigger_word) and len(result) > len(self.trigger_word) + 1:
```

Comment:
Again, I disagree with this. The reason for having the list of command words is to make sure that always the word from that list is recognized, or nothing is. E.g. is the user says "grasp" but the recognizer gets it as "grass", we still want to execute grasp command, instead of sending a "grass" word to the topic which will be unused. That's why the program, whichever word is recognized after the trigger word, should approximate it to the closesed word in the command words list. Otherwise, we loose a lot in terms of robustness and usability as far as I'm concerned. I we really want to preserve ability to have arbitrary commands, we can always have an option to pass empty list of commands, and that way tell the program to use arbitrary ones.

Reply:
I think it is more practical to move usage specific post processing to the places where it is being actually used. Otherwise if multiple nodes need voice control they would need to span multiple speech recognition nodes with they own command lists and would waste resources by sending the same audio to Google multiple times. Now this node has a single purpose - speech recognition. Other nodes are free to post process recognised command to their needs.

Reply:
But this is speech control, not just speech recognition and it's purpuse is to send commands, not voice recognized words/sentences. If we move post processing outside we have to possibly reproduce the code for it, and maintain it multiple times, also it breaks the SRP rule. If we ever need to have mutilple nodes controlled by this, we can always have commands for one and another combined in one list passed to this node. We get the benefit that you describe while maintaining robustness and cleanness of this node. This might be just my opinion, happy to discuss in general and let it go if more people disagree with my approach.

Reply:
Also, we can always have the best of all worlds by allowing your way when passing empty list of command words to the class for example

Reply:
If for launching speech recognition node you would need to know all commands used by any other nodes, how you could call it compliant with single-responsibility principle? This would mean that speech recognition node depends on other nodes (need to know about) and other nodes depend on it. Potentially such list could go to some central configuration server providing configuration for whole system. My point is not only where to keep command list though. My worry is that `get_close_matches` function might be not good enough for all potential usages. For more critical operations risk of false positives might be too big to better repeat command than get it executed at a wrong time. Also if `get_close_matches` is used centrally in this node then there might be a lot of similar commands in global pool of commands, but commands used by single node can be chosen to be dissimilar without thinking how it would affect other nodes.

Reply:
The node is node dependant on other nodes, you can call launch file from a shared launch file and pass commands there and node implementation doesnt need to know about it. Also, again, we can always just leave empty list if needed.
I agree with the get_close_matches concert, but my bigger worry is that this node immediately looses its usability as a stand-alone send-commands tool, since there will be far too many wrongly recognised words. For any of its clients now, you always have to remember to implement correct filtering.

can we comprimise by leaving a method for filtering here and use it as a parameter? e.g. only use it if "filer_to_command_words" or similar is set to true? That way we can set it to false if there is a sensitive client or we just want custom filtering? And leave the ability to pass non-empty command words? That way we can maintain flexibility you talk about but still use this as stand-alone, robust node if desired.

Reply:
Re-added auto correction functionality as ROS parameter in db72bb37f526ecc2da77663c849eaa046ace07b2.